cogs/strawpoll.py

```diff
@@ -0,0 +1,292 @@
+        self.bot = bot
+        self.settings_path = settings_path
+        self.poll_sessions = []
+        self.poll_session_tasks = {}
+        try:
+            if not dataIO.is_valid_json(settings_path):
+                raise FileNotFoundError("JSON file not valid")
+            settings = dataIO.load_json(settings_path)
+            self.settings = defaultdict(dict, settings)
+        except FileNotFoundError:
+            self.settings = {}
+            if not os.path.exists(os.path.dirname(settings_path)):
+                print("Creating strawpoll data folder...")
+                os.makedirs(os.path.dirname(settings_path))
+            dataIO.save_json(settings_path, self.settings)
+
+    def _new_server_settings(self, server_id):
+        default_settings = {
+            'refresh_emoji': '🔄',
+            'poll_length': 60.0,
+            'poll_react_time': 1800.0, # 30 minutes
+            'bar_length': 30
+        }
+        self.settings[server_id] = default_settings
+        dataIO.save_json(self.settings_path, self.settings)
+
+    def _check_new_poll_tasks(self):
+        for poll in self.poll_sessions:
+            if poll not in self.poll_session_tasks:
+                self.poll_session_tasks[poll] = asyncio.ensure_future(
+                    self._check_reacts(poll))
+
+    def _cancel_expired_poll_tasks(self, current_time):
+        for poll in self.poll_session_tasks.keys():
+            server_id = poll.message.server.id
+            poll_react_time = self.settings[server_id]['poll_react_time']
+            if current_time-poll.start_time > poll_react_time:
+                self.poll_session_tasks[poll].cancel()
+ 
+    def _delete_expired_poll_sessions(self, current_time):
+        trimmed_poll_sessions = []
+        for poll in self.poll_sessions:
+            server_id = poll.message.server.id
+            poll_react_time = self.settings[server_id]['poll_react_time']
+            if current_time-poll.start_time < poll_react_time:
+                trimmed_poll_sessions.append(poll)
+        self.poll_sessions = trimmed_poll_sessions
+
+    def _remove_done_poll_tasks(self):
+        self.poll_session_tasks = { 
+            poll:task for poll,task in self.poll_session_tasks.items() 
+            if not task.done()}
+
+    async def check_polls(self):
+        while self is self.bot.get_cog("Strawpoll"):
+            current_time = time()
+            self._check_new_poll_tasks()
+            self._cancel_expired_poll_tasks(current_time)
+            self._delete_expired_poll_sessions(current_time)
+            self._remove_done_poll_tasks()
+            await asyncio.sleep(5)
+        for poll in self.poll_session_tasks.keys():
+            self.poll_session_tasks[poll].cancel()
+        
+    async def _check_reacts(self, poll):
+        settings = self.settings[poll.message.server.id]
+        refresh_emoji = settings['refresh_emoji']
+        try:
+            while True:
```

Comment:
I'd like to go over this infinite loop, I think it might be possible to reorganize it

Reply:
It being in a try block waiting to be interrupted with a cancel is why I was comfortable with making it an infinite loop, but I'd be down to talk about it and explore other ways to do this check routine